File: laba4/task4.go
package main

import (
	"fmt"
)

// input считывает данные о кандидатах и голосах групп
func input(n, m int) ([]string, []int, [][]int) {
	candidates := make([]string, n)
	voters := make([]int, m)
	ranking := make([][]int, m)

	fmt.Println("Введите имена кандидатов:")
	for i := 0; i < n; i++ {
		fmt.Scan(&candidates[i])
	}

	fmt.Println("Для каждой группы сначала количество избирателей,")
	fmt.Println("затем порядок кандидатов (от лучшего к худшему, индексы от 1 до n):")

	for i := 0; i < m; i++ {
		fmt.Printf("Группа %d:\n", i+1)
		fmt.Scan(&voters[i])

		ranking[i] = make([]int, n)
		for j := 0; j < n; j++ {
			var val int
			fmt.Scan(&val)
			ranking[i][j] = val - 1 // приводим к 0-индексации
		}
	}

	return candidates, voters, ranking
}

func main() {
	var n, m int

	fmt.Print("Количество кандидатов: ")
	fmt.Scan(&n)

	fmt.Print("Количество групп избирателей: ")
	fmt.Scan(&m)

	candidates, voters, ranking := input(n, m)

	// --- Метод Борда ---
	borda := make([]int, n)
	for i := 0; i < m; i++ { // по группам
		for pos := 0; pos < n; pos++ { // по позициям
			candidate := ranking[i][pos]
			// Очки: (n - позиция - 1). Первое место (pos=0) дает n-1 очков.
			borda[candidate] += (n - pos - 1) * voters[i]
		}
	}

	bordaWinner := 0
	for i := 1; i < n; i++ {
		if borda[i] > borda[bordaWinner] {
			bordaWinner = i
		}
	}

	// --- Метод Кондорсе ---
	condorcetWinner := -1

	for a := 0; a < n; a++ {
		winsAll := true // гипотеза: кандидат 'a' побеждает всех

		for b := 0; b < n; b++ {
			if a == b {
				continue
			}

			votesA := 0
			votesB := 0

			for g := 0; g < m; g++ {
				posA, posB := 0, 0
				for k := 0; k < n; k++ {
					if ranking[g][k] == a {
						posA = k
					}
					if ranking[g][k] == b {
						posB = k
					}
				}

				if posA < posB {
					votesA += voters[g]
				} else {
					votesB += voters[g]
				}
			}

			// Если в паре (a, b) 'a' не набрал строго больше голосов, он не Кондорсе-победитель
			if votesA <= votesB {
				winsAll = false
				break
			}
		}

		if winsAll {
			condorcetWinner = a
			break
		}
	}

	// --- Вывод результатов ---
	fmt.Println("\nМетод Борда:")
	for i := 0; i < n; i++ {
		fmt.Printf("%s: %d очков\n", candidates[i], borda[i])
	}
	fmt.Printf("Победитель: %s (%d очков)\n", candidates[bordaWinner], borda[bordaWinner])

	fmt.Println("\nМетод Кондорсе:")
	if condorcetWinner == -1 {
		fmt.Println("Победитель не определён (парадокс Кондорсе)")
	} else {
		fmt.Printf("Победитель: %s\n", candidates[condorcetWinner])
	}
}

File: laba4/task4.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

void input(int& n, int& m, vector<string>& candidates, vector<int>& voters, vector<vector<int>>& ranking){

    cout << "Введите имена кандидатов"<<endl;
    for (int i = 0; i < n; i++) {
        cin >> candidates[i];
    }
    
    cout << "Для каждой группы сначала количество избирателей"<<endl;
    cout << "затем порядок кандидатов (от лучшего к худшему)"<<endl;

    for (int i = 0; i < m; i++) {
        cout << "Группа " << i + 1 << ":\n";

        cin >> voters[i]; // количество избирателей в группе

        for (int j = 0; j < n; j++) { // порядок кандидатов
            cin >> ranking[i][j];
            ranking[i][j]--; // чтобы начаналось с 0
        }
    }

}

int main() {
    int n, m;

    cout << "Количество кандидатов ";
    cin >> n;

    cout << "Количество групп избирателей ";
    cin >> m;

    vector<string> candidates(n);
    vector<int> voters(m);
    vector<vector<int>> ranking(m, vector<int>(n));
    input(n, m, candidates, voters, ranking);
    

    vector<int> borda(n, 0); // очки кандидатов
    
    for (int i = 0; i < m; i++) { // проходимся по группам
        for (int pos = 0; pos < n; pos++) { // проходимся по позициям кандидатам
            int candidate = ranking[i][pos];
            borda[candidate] += (n - pos - 1) * voters[i]; // начисление очков 
        }
    }

    int bordaWinner = 0;

    for (int i = 1; i < n; i++) { // ищем кандидата с максимальным числом очков
        if (borda[i] > borda[bordaWinner]) {
            bordaWinner = i;
        }
    }



    int condorcetWinner = -1;

    for (int a = 0; a < n; a++) {
        bool winsAll = true; // считаем, что кандидат побеждает всех

        for (int b = 0; b < n; b++) {
            if (a == b) continue;

            int votesA = 0;
            int votesB = 0;

            for (int g = 0; g < m; g++) {

                int posA = 0, posB = 0;

                for (int k = 0; k < n; k++) {
                    if (ranking[g][k] == a) {
                        posA = k;
                    }
                    if (ranking[g][k] == b) {
                        posB = k;
                    }
                }

                if (posA < posB)
                    votesA += voters[g];
                else
                    votesB += voters[g];
            }

            if (votesA <= votesB) {
                winsAll = false;
                break;
            }
        }

        if (winsAll) {
            condorcetWinner = a;
            break;
        }
    }


    cout << endl<<"Метод Борда "<<endl;

    for (int i = 0; i < n; i++) {
        cout << candidates[i] << ": " << borda[i] << " очков"<<endl;
    }

    cout << "Победитель: " << candidates[bordaWinner] <<" очков "<< borda[bordaWinner] << endl;

    cout <<endl<< "Метод Кондорсе"<<endl;

    if (condorcetWinner == -1)
        cout << "Победитель не определён"<<endl;
    else
        cout << "Победитель: " << candidates[condorcetWinner] << endl;

    return 0;
}